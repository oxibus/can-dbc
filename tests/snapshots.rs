#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
