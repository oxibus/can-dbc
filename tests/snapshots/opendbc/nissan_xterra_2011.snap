---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - XXX
messages:
  - id:
      Standard: 2
    name: STEERING
    size: 5
    transmitter:
      NodeName: XXX
    signals:
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 35
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: STEERING_ANGLE
        multiplexer_indicator: Plain
        start_bit: 0
        size: 16
        byte_order: LittleEndian
        value_type: Signed
        factor: 0.1
        offset: 0
        min: 0
        max: 65535
        unit: deg
        receivers:
          - XXX
      - name: POWER_STEER_RATE
        multiplexer_indicator: Plain
        start_bit: 23
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 505
    name: ENGINE_1
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: RPM
        multiplexer_indicator: Plain
        start_bit: 16
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.125
        offset: 0
        min: 0
        max: 45000
        unit: R/min
        receivers:
          - XXX
      - name: FAN_REQ
        multiplexer_indicator: Plain
        start_bit: 6
        size: 2
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: AC_REQ
        multiplexer_indicator: Plain
        start_bit: 3
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 561
    name: ENGINE_2
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: Pedal_position
        multiplexer_indicator: Plain
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.5
        offset: 0
        min: 0
        max: 200
        unit: "%"
        receivers:
          - XXX
  - id:
      Standard: 563
    name: ENGINE_7
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CLT
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.366666
        offset: 0
        min: 0
        max: 255
        unit: C
        receivers:
          - XXX
      - name: RPMlow
        multiplexer_indicator: Plain
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 3.15
        offset: 0
        min: 0
        max: 45000
        unit: R/min
        receivers:
          - XXX
      - name: RPMhi
        multiplexer_indicator: Plain
        start_bit: 56
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 3.15
        offset: 0
        min: 0
        max: 45000
        unit: R/min
        receivers:
          - XXX
  - id:
      Standard: 573
    name: ENGINE_3
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: Pedal_position
        multiplexer_indicator: Plain
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.392
        offset: 0
        min: 0
        max: 255
        unit: "%"
        receivers:
          - XXX
      - name: Throttle_position_capped
        multiplexer_indicator: Plain
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.392
        offset: 0
        min: 0
        max: 255
        unit: "%"
        receivers:
          - XXX
      - name: RPM
        multiplexer_indicator: Plain
        start_bit: 24
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 3.15
        offset: 0
        min: 0
        max: 45000
        unit: R/min
        receivers:
          - XXX
      - name: CLT
        multiplexer_indicator: Plain
        start_bit: 56
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.366666
        offset: 0
        min: 0
        max: 255
        unit: C
        receivers:
          - XXX
  - id:
      Standard: 574
    name: ENGINE_4
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: Throttle_position_inverted
        multiplexer_indicator: Plain
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.392
        offset: 0
        min: 0
        max: 255
        unit: "%"
        receivers:
          - XXX
      - name: EstimatedTorque
        multiplexer_indicator: Plain
        start_bit: 24
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 45000
        unit: nM
        receivers:
          - XXX
      - name: Throttle_position
        multiplexer_indicator: Plain
        start_bit: 48
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.392
        offset: 0
        min: 0
        max: 255
        unit: "%"
        receivers:
          - XXX
  - id:
      Standard: 595
    name: TCU_1
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: SHAFT_1_SPEED
        multiplexer_indicator: Plain
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 45000
        unit: r/min
        receivers:
          - XXX
      - name: SHAFT_2_SPEED
        multiplexer_indicator: Plain
        start_bit: 48
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 45000
        unit: r/min
        receivers:
          - XXX
  - id:
      Standard: 640
    name: SPEED
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: SPEED
        multiplexer_indicator: Plain
        start_bit: 32
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 45000
        unit: km/h
        receivers:
          - XXX
  - id:
      Standard: 644
    name: ABS_1
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: WHEEL_1
        multiplexer_indicator: Plain
        start_bit: 0
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 45000
        unit: km/h
        receivers:
          - XXX
      - name: WHEEL_2
        multiplexer_indicator: Plain
        start_bit: 15
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 45000
        unit: km/h
        receivers:
          - XXX
  - id:
      Standard: 645
    name: ABS_2
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: WHEEL_3
        multiplexer_indicator: Plain
        start_bit: 0
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 45000
        unit: km/h
        receivers:
          - XXX
      - name: WHEEL_4
        multiplexer_indicator: Plain
        start_bit: 15
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 45000
        unit: km/h
        receivers:
          - XXX
  - id:
      Standard: 852
    name: SPEED_BREAK
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: BREAK_LIGHT
        multiplexer_indicator: Plain
        start_bit: 52
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: SPEED_MPH
        multiplexer_indicator: Plain
        start_bit: 7
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.0066
        offset: 0
        min: 0
        max: 65535
        unit: mph
        receivers:
          - XXX
      - name: TCS_OFF
        multiplexer_indicator: Plain
        start_bit: 38
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 861
    name: BCM
    size: 8
    transmitter:
      NodeName: XXX
  - id:
      Standard: 1361
    name: ENGINE_5
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CLT
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: -45
        min: -45
        max: 150
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 1408
    name: ENGINE_6
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: RPM
        multiplexer_indicator: Plain
        start_bit: 7
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1.5
        offset: 0
        min: 0
        max: 65535
        unit: ""
        receivers:
          - XXX
      - name: OIL_TEMP
        multiplexer_indicator: Plain
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: -40
        min: -40
        max: 150
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 1477
    name: MILEAGE
    size: 8
    transmitter:
      NodeName: XXX
